### PR TITLE
SR-11283: Amend OptionSet.remove method

### PR DESCRIPTION
**Pull request by JacobMao**
This PR modify OptionSet.remove method to make it conform to the description of the document.
Currently this method returns wrong result when element intersects but is not fully contained.
I also added some test cases for this.

Resolves [SR-11283.](https://bugs.swift.org/browse/SR-11283)

### CODE REVIEW

**Comment by JacobMao** [comment]
cc @lilyball

**Comment by JacobMao** [comment]
@swift-ci test

File: stdlib/public/core/OptionSet.swift
Comment (lilyball) [inline stdlib/public/core/OptionSet.swift]: If `intersectionElements.isEmpty` then surely we can skip the subtraction, no?

Comment (JacobMao) [reply stdlib/public/core/OptionSet.swift]: Yes.
But I prefer to keep code clean and don't add this condition statement at here, and follow other methods of OptionSet

**Comment by CodaFi** [comment]
@JacobMao Only those with the commit bit are allowed to trigger PR testing for security reasons.

Could you please squash the merge commits out of this pull request and just leave behind a single commit with your changes?

**Comment by JacobMao** [comment]
@CodaFi 
thanks for your review. sorry, I forgot to squash those commits.
I will do it immediately

**Comment by CodaFi** [comment]
I think it's still not quite right.  You can run `git rebase` (I prefer an interactive one with `-i`) and delete these commits.

File: stdlib/public/core/OptionSet.swift
Comment (lorentey) [reply stdlib/public/core/OptionSet.swift]: I agree with Lily -- an early exit with a `guard` would read best here.

**Review by lorentey — APPROVED** [review]
Thank you! I requested a slight coding style change, but otherwise this looks ready to land.

I'll be able to squash these commits when I merge this -- don't worry too much about removing them.

**Comment by lorentey** [comment]
@swift-ci

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
Note: To help us manage compatibility risks, we'll need to delay merging this PR until the 5.2 branch is cut. (Currently [scheduled for December 9th][branching].)

[branching]: https://forums.swift.org/t/swift-5-2-branching-date-re-scheduled-for-december-9th/30478

File: stdlib/public/core/OptionSet.swift
Comment (JacobMao) [reply stdlib/public/core/OptionSet.swift]: @lorentey 
Thanks for your review. I have changed code at [d5e2533](https://github.com/apple/swift/pull/28378/commits/d5e2533ceff38815c15f89a44048393d501ce439)

**Comment by JacobMao** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test macOS platform

**Comment by CodaFi** [comment]
This seems like an infrastructure problem.  Maybe the smoke test bots are having a better time

@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
Ugh, it's still happening.  Ignore the macOS build.  We'll get that sorted tomorrow.

**Comment by CodaFi** [comment]
@swift-ci please clean test macOS platform

**Comment by CodaFi** [comment]
Ah, this picked up Doug's failure from this morning.  Let's try one more time.

@swift-ci please clean test macOS platform

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
Oh, alas, I forgot to squash the commits. Sorry about that!

**Comment by shahmishal** [comment]
We are seeing test failure after merging this PR:

```
Swift(macosx-x86_64) :: stdlib/OptionSetTest.swift
```

**Comment by shahmishal** [comment]
```
20:36:47 ******************** TEST 'Swift(macosx-x86_64) :: stdlib/OptionSetTest.swift' FAILED ********************
20:36:47 Script:
20:36:47 --
20:36:47 : 'RUN: at line 12';   rm -rf "/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/test-macosx-x86_64/stdlib/Output/OptionSetTest.swift.tmp" && mkdir -p "/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/test-macosx-x86_64/stdlib/Output/OptionSetTest.swift.tmp" && xcrun --toolchain default --sdk '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk' /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/bin/swiftc -target x86_64-apple-macosx10.9  -module-cache-path '/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache' -F '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks' -toolchain-stdlib-rpath -Xlinker -rpath -Xlinker '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks'  -swift-version 4  -Xfrontend -ignore-module-source-info  -F '/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/lib' -Xlinker -rpath -Xlinker '/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/lib' -module-cache-path '/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache' /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/swift/test/stdlib/OptionSetTest.swift -o /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/test-macosx-x86_64/stdlib/Output/OptionSetTest.swift.tmp/a.out -module-name main  && codesign -f -s - /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/test-macosx-x86_64/stdlib/Output/OptionSetTest.swift.tmp/a.out &&/usr/bin/env DYLD_LIBRARY_PATH='/usr/lib/swift:/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/lib/swift/macosx' LD_LIBRARY_PATH='/usr/lib/swift:/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/lib/swift/macosx:' SIMCTL_CHILD_DYLD_LIBRARY_PATH='/usr/lib/swift:/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/lib/swift/macosx'  /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/test-macosx-x86_64/stdlib/Output/OptionSetTest.swift.tmp/a.out
20:36:47 --
20:36:47 Exit Code: 134
20:36:47 
20:36:47 Command Output (stdout):
20:36:47 --
20:36:47 [ RUN      ] OptionSet.basics
20:36:47 [       OK ] OptionSet.basics
20:36:47 [ RUN      ] OptionSet.set algebra
20:36:47 stdout>>> check failed at /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/swift/test/stdlib/OptionSetTest.swift, line 98
20:36:47 stdout>>> expected: Optional(main.PackagingOptions(rawValue: 4)) (of type Swift.Optional<main.PackagingOptions>)
20:36:47 stdout>>> actual: nil (of type Swift.Optional<main.PackagingOptions>)
20:36:47 [     FAIL ] OptionSet.set algebra
20:36:47 OptionSet: Some tests failed, aborting
20:36:47 UXPASS: []
20:36:47 FAIL: ["set algebra"]
20:36:47 SKIP: []
20:36:47 To debug, run:
20:36:47 $ /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/test-macosx-x86_64/stdlib/Output/OptionSetTest.swift.tmp/a.out --stdlib-unittest-in-process --stdlib-unittest-filter "set algebra"
20:36:47 
20:36:47 --
20:36:47 Command Output (stderr):
20:36:47 --
20:36:47 /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/test-macosx-x86_64/stdlib/Output/OptionSetTest.swift.script: line 1: 55283 Abort trap: 6           /usr/bin/env DYLD_LIBRARY_PATH='/usr/lib/swift:/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/lib/swift/macosx' LD_LIBRARY_PATH='/usr/lib/swift:/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/lib/swift/macosx:' SIMCTL_CHILD_DYLD_LIBRARY_PATH='/usr/lib/swift:/Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/lib/swift/macosx' /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/buildbot-custom/swift-macosx-x86_64/test-macosx-x86_64/stdlib/Output/OptionSetTest.swift.tmp/a.out
20:36:47 
20:36:47 --
20:36:47 
20:36:47 ********************
```

**Comment by lorentey** [comment]
Ah, we need to add availability checks to the new tests, or they will keep failing on back deployment tests. I'll take care of it.

**Pull request merged**
